### PR TITLE
Create monte_carlo_integration_univariate.py

### PR DESCRIPTION
**Pull request by SatyakiMandal**
Added integration/area under the curve using the Monte Carlo method

### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by SatyakiMandal** [comment]
Review required

**Review by realstealthninja — APPROVED** [review]
lgtm

File: maths/monte_carlo_integration_univariate.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    plt.hist(plt_vals, bins=30, ec="black")
+
+    # sets the label of the x-axis of the plot
+    plt.xlabel("Areas")
+    plt.show() # shows the plot
+
+    #-----END OF PLOT SECTION (OPTIONAL)------#
+    """
+
+    return sum(plt_vals) / number_of_sections  # takinf the average value
```
Comment (realstealthninja) [inline maths/monte_carlo_integration_univariate.py:94]: taking

Comment (SatyakiMandal) [reply maths/monte_carlo_integration_univariate.py]: Corrected

File: maths/monte_carlo_integration_univariate.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    # Doctest
+    """
+    :param univariate_variable: int
+    :return: float
+
+    >>> round(function_to_be_integrated(0))
+    0
+    """
+
+    return np.sin(univariate_variable)  # example function
```
Comment (ZeroDayOwl) [inline maths/monte_carlo_integration_univariate.py:31]: This is a trivial comment, you can remove it.

File: maths/monte_carlo_integration_univariate.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        # values of univariate_variable
+        integral = 0.0
+
+        # iterates and sums up values of different functions
+        # of univariate_variable
+        for i in ar:
+            integral += function_to_be_integrated(i)
+
+        # we get the answer by the formula derived adobe
+        ans = (upper_limit - lower_limit) / float(number_of_sections) * integral
```
Comment (ZeroDayOwl) [inline maths/monte_carlo_integration_univariate.py:73]:
```suggestion
        answer = (upper_limit - lower_limit) / float(number_of_sections) * integral
```

File: maths/monte_carlo_integration_univariate.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        # appends the solution to a list for plotting the graph
+        plt_vals.append(ans)
```
Comment (ZeroDayOwl) [inline maths/monte_carlo_integration_univariate.py:75]:
```suggestion
        plt_vals.append(answer)
```

File: maths/monte_carlo_integration_univariate.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+# example limits
+lower_limit = 0
+upper_limit = np.pi  # gets the value of pi
+
+number_of_sections = 1000  # Number of individual ares to be considered
+
+# function call
+# the final area under the curve(integration) value is considered as the average
+# of all the individual areas calculated
+print(f"Approx. value: {monte_carlo(lower_limit,upper_limit,number_of_sections)}.")
```
Comment (ZeroDayOwl) [inline maths/monte_carlo_integration_univariate.py:111]: You should write a doctest to show how the entire algorithm would execute instead of this example piece of code.

**Review by ZeroDayOwl — CHANGES_REQUESTED** [review]

File: maths/monte_carlo_integration_univariate.py (hunk 5 above)
Comment (SatyakiMandal) [reply maths/monte_carlo_integration_univariate.py:111]: Since, doctests have been added to individual the functions. Would it be necessary to  add a doctest here or should an explanatory comment suffice?

**Comment by SatyakiMandal** [comment]
Please add a review or merge.

File: maths/monte_carlo_integration_univariate.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
+"""
+This program finds the approximate integration value/area under the curve of
+a specified function within specified limits using Monte Carlo integration method.
+
+Further, a graph of the individual areas under the curve considered for the calculation
+is also plotted. (PLOT SECTION -> Optional implementation)
+"""
```
Comment (realstealthninja) [inline maths/monte_carlo_integration_univariate.py:7]: Adding a link to the source might be helpful

File: maths/monte_carlo_integration_univariate.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+import doctest
+
+# importing the modules
+import random
+
+# import matplotlib.pyplot as plt
```
Comment (realstealthninja) [inline maths/monte_carlo_integration_univariate.py:14]:
```suggestion
```

File: maths/monte_carlo_integration_univariate.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+This program finds the approximate integration value/area under the curve of
+a specified function within specified limits using Monte Carlo integration method.
+
+Further, a graph of the individual areas under the curve considered for the calculation
+is also plotted. (PLOT SECTION -> Optional implementation)
+"""
+
+import doctest
+
+# importing the modules
```
Comment (realstealthninja) [inline maths/monte_carlo_integration_univariate.py:11]:
```suggestion
```

File: maths/monte_carlo_integration_univariate.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
+"""
+This program finds the approximate integration value/area under the curve of
+a specified function within specified limits using Monte Carlo integration method.
+
+Further, a graph of the individual areas under the curve considered for the calculation
+is also plotted. (PLOT SECTION -> Optional implementation)
+"""
+
+import doctest
+
```
Comment (realstealthninja) [inline maths/monte_carlo_integration_univariate.py:10]:
```suggestion
```

File: maths/monte_carlo_integration_univariate.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+# importing the modules
+import random
+
+# import matplotlib.pyplot as plt
+import numpy as np
+
+
+# function to calculate the sin of a particular value of x
```
Comment (realstealthninja) [inline maths/monte_carlo_integration_univariate.py:18]: Add this to the docstring instead of a comment

File: maths/monte_carlo_integration_univariate.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+# define your function
```
Comment (realstealthninja) [inline maths/monte_carlo_integration_univariate.py:19]:
```suggestion
```

File: maths/monte_carlo_integration_univariate.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+    plt.xlabel("Areas")
+    plt.show() # shows the plot
+
+    #-----END OF PLOT SECTION (OPTIONAL)------#
+    """
+
+    return sum(plt_vals) / number_of_sections  # taking the average value
+
+
+doctest.testmod()
```
Comment (realstealthninja) [inline maths/monte_carlo_integration_univariate.py:97]:
```suggestion
if __name__ == "__main__":
     doctest.testmod()
```

File: maths/monte_carlo_integration_univariate.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+# example limits
+lower_limit = 0
+upper_limit = np.pi  # gets the value of pi
+
+number_of_sections = 1000  # Number of individual ares to be considered
+
+# function call
+# the final area under the curve(integration) value is considered as the average
+# of all the individual areas calculated
+print(f"Approx. value: {monte_carlo(lower_limit,upper_limit,number_of_sections)}.")
```
Comment (realstealthninja) [inline maths/monte_carlo_integration_univariate.py:111]: Why is this here?

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: maths/monte_carlo_integration_univariate.py (hunk 13 above)
Comment (SatyakiMandal) [reply maths/monte_carlo_integration_univariate.py:108]: Defining the limits of integration (lower_limit, upper_limit); number of sections, the integration is divided into; and the method call, which is printed.

File: maths/monte_carlo_integration_univariate.py (hunk 7 above)
Comment (SatyakiMandal) [reply maths/monte_carlo_integration_univariate.py:12]: The import is necessary for the plotting section, which has been kept as an optional feature for the user.

File: maths/monte_carlo_integration_univariate.py [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+
+import doctest
+import random
+
+import numpy as np
+
+
+def function_to_be_integrated(univariate_variable: int) -> float:
+
+    # Doctest
```
Comment (CaedenPH) [inline maths/monte_carlo_integration_univariate.py:21]:
```suggestion
```

File: maths/monte_carlo_integration_univariate.py [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+    """
+    Function to calculate the sin of a particular value of x
+
+    :param univariate_variable: int
+    :return: float
+
+    >>> round(function_to_be_integrated(0))
+    0
+    """
+
```
Comment (CaedenPH) [inline maths/monte_carlo_integration_univariate.py:31]:
```suggestion
```

File: maths/monte_carlo_integration_univariate.py [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+def function_to_be_integrated(univariate_variable: int) -> float:
+
+    # Doctest
+    """
+    Function to calculate the sin of a particular value of x
+
+    :param univariate_variable: int
+    :return: float
+
```
Comment (CaedenPH) [inline maths/monte_carlo_integration_univariate.py:27]:
```suggestion
```
This doesn't tell us anything that we don't already know

File: maths/monte_carlo_integration_univariate.py [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+    >>> round(function_to_be_integrated(0))
+    0
+    """
+
+    return np.sin(univariate_variable)
+
+
+def monte_carlo(lower_limit: int, upper_limit: int, number_of_sections: int) -> float:
+
+    # Doctest
```
Comment (CaedenPH) [inline maths/monte_carlo_integration_univariate.py:37]:
```suggestion
```

File: maths/monte_carlo_integration_univariate.py [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+    """
+    Monte Carlo integration function
+
+    :param lower_limit: int
+    :param upper_limit: int
+    :param N: int
+    :return: float
+
```
Comment (CaedenPH) [inline maths/monte_carlo_integration_univariate.py:45]:
```suggestion
```
This doesn't tell us anything we don't already know

File: maths/monte_carlo_integration_univariate.py [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+    >>> round(monte_carlo(0, np.pi, 1000))
```
Comment (CaedenPH) [inline maths/monte_carlo_integration_univariate.py:46]: Perhaps you could write some more doctests

Comment (SatyakiMandal) [reply maths/monte_carlo_integration_univariate.py:33]: added a few more doctests

File: maths/monte_carlo_integration_univariate.py [hunk 20]
```diff
@@ -0,0 +1,111 @@
+"""
+This program finds the approximate integration value/area under the curve of
+a specified function within specified limits using Monte Carlo integration method.
+
+Further, a graph of the individual areas under the curve considered for the calculation
+is also plotted. (PLOT SECTION -> Optional implementation)
+
+Link to source material:
+https://towardsdatascience.com/the-basics-of-monte-carlo-integration-5fe16b40482d
+
```
Comment (CaedenPH) [inline maths/monte_carlo_integration_univariate.py:10]:
```suggestion
```

**Review by CaedenPH — COMMENTED** [review]
@cclauss Can we get a final review from you? Looks good to me

File: maths/monte_carlo_integration_univariate.py [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+        for i in ar:
+            integral += function_to_be_integrated(i)
+
+        # we get the answer by the formula derived adobe
+        answer = (upper_limit - lower_limit) / float(number_of_sections) * integral
+        # appends the solution to a list for plotting the graph
+        plt_vals.append(answer)
+
+    """
+    #--------PLOT SECTION (OPTIONAL)----------#
```
Comment (cclauss) [inline maths/monte_carlo_integration_univariate.py:75]: Please create a separate plot function.

File: maths/monte_carlo_integration_univariate.py (hunk 19 above)
Comment (cclauss) [reply maths/monte_carlo_integration_univariate.py:32]: Please add doctests where lower_limit is negative, a float, a string.  Same for upper_limit and number_of_sections.  Show that the function will raise Exceptionswhen handed garbage input.

File: maths/monte_carlo_integration_univariate.py (hunk 21 above)
Comment (SatyakiMandal) [reply maths/monte_carlo_integration_univariate.py]: Added a separate plot function

File: maths/monte_carlo_integration_univariate.py (hunk 19 above)
Comment (SatyakiMandal) [reply maths/monte_carlo_integration_univariate.py:34]: Required doctests and additional changes have been added

File: maths/monte_carlo_integration_univariate.py [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+    2
+    >>> round(monte_carlo(0, np.pi, 10))
+    2
+    >>> round(monte_carlo(0, 2*np.pi, 1000))
+    0
+    >>> round(monte_carlo(-2*np.pi, 2*np.pi, 1000))
+    0
+    >>> round(monte_carlo(-10.2345678, 1.89712, 1000))
+    0
+    """
```
Comment (ZeroDayOwl) [inline maths/monte_carlo_integration_univariate.py:48]: You haven't added exceptional cases where the function would raise Value error for garbage input as @cclauss mentioned. Add those invalid cases as well.

File: maths/monte_carlo_integration_univariate.py [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+
+import numpy as np
+
+
+def function_to_be_integrated(univariate_variable: float) -> float:
+    """
+    Function to calculate the sin of a particular value of x
+
+    >>> round(function_to_be_integrated(0))
+    0
```
Comment (ZeroDayOwl) [inline maths/monte_carlo_integration_univariate.py:23]: Add some invalid cases here as well.

**Review by ZeroDayOwl — CHANGES_REQUESTED** [review]
